### PR TITLE
nxos_vrf fix

### LINKED ISSUE #36461: nxos_vrf error when using vni option
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

nxos_vrf

##### ANSIBLE VERSION

```
ansible 2.6.0
  config file = /Users/jdacosta/Devel/ansible-netops/cisco/nxos/applications/fabric/europe-central-fr1/ansible.cfg
  configured module search path = [u'/Users/jdacosta/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.10 (default, Feb  7 2017, 00:08:15) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```

##### OS / ENVIRONMENT

N/A
nxos 7.0(3)I7(1)

##### STEPS TO REPRODUCE

```yaml
- name: "Ensure state of VRF for VNIs"
  nxos_vrf:
    vrf: "tenant01"
    rd:  auto
    vni: "10150"
    state: "present"
    admin_state: "up"

```

##### ACTUAL RESULTS

traceback error

```
TASK [victorock.overlay-nxos-leaf : Ensure state of VRF for VNIs] ************************************************************
task path: /Users/jdacosta/Devel/ansible-netops/cisco/nxos/applications/fabric/europe-central-fr1/roles/victorock.overlay-nxos-leaf/tasks/vni.yml:14
<127.0.0.1> attempting to start connection
<127.0.0.1> using connection plugin network_cli
<127.0.0.1> found existing local domain socket, using it!
<127.0.0.1> updating play_context for connection
<127.0.0.1> 
<127.0.0.1> local domain socket path is /Users/jdacosta/.ansible/pc/32fbfb2e00
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/network/__init__.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/basic.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/network/nxos/__init__.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/network/common/utils.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/network/common/__init__.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/network/nxos/nxos.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/_text.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/parsing/__init__.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/pycompat24.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/six/__init__.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/connection.py
Using module_utils file /Library/Python/2.7/site-packages/ansible/module_utils/urls.py
Using module file /Library/Python/2.7/site-packages/ansible/modules/network/nxos/nxos_vrf.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: jdacosta
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571 `" && echo ansible-tmp-1519162691.84-261532122578571="` echo /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571 `" ) && sleep 0'
<127.0.0.1> PUT /Users/jdacosta/.ansible/tmp/ansible-local-27233QmXSr2/tmp2bG1bG TO /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571/nxos_vrf.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571/ /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571/nxos_vrf.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571/nxos_vrf.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/jdacosta/.ansible/tmp/ansible-tmp-1519162691.84-261532122578571/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/jv/29pmt_2j39ngz8ftz40j1mn00000gn/T/ansible_Tj2nE6/ansible_module_nxos_vrf.py", line 467, in <module>
    main()
  File "/var/folders/jv/29pmt_2j39ngz8ftz40j1mn00000gn/T/ansible_Tj2nE6/ansible_module_nxos_vrf.py", line 453, in main
    commands = map_obj_to_commands((want, have), module)
  File "/var/folders/jv/29pmt_2j39ngz8ftz40j1mn00000gn/T/ansible_Tj2nE6/ansible_module_nxos_vrf.py", line 258, in map_obj_to_commands
    if have.get('vni') and have.get('vni') != '':
AttributeError: 'list' object has no attribute 'get'

failed: [leaf01] (item=nxos_vrf_vni tenant01) => {
    "changed": false, 
    "failed": true, 
    "nxos_vrf_vni": {
        "key": "l2_150_10150", 
        "value": {
            "multicast": {
                "group": "239.0.0.150"
            }, 
            "name": "l2_150_10150", 
            "route": {
                "distinguisher": "auto", 
                "target_both": "auto"
            }, 
            "states": {
                "admin": "up", 
                "feature": "enabled", 
                "shutdown": false, 
                "state": "present"
            }, 
            "svi": "10.1.150.1/24", 
            "vlan": 150, 
            "vni": 10150, 
            "vrf": "tenant01"
        }
    }, 
    "rc": 1
}

MSG:

MODULE FAILURE


MODULE_STDERR:

Traceback (most recent call last):
  File "/var/folders/jv/29pmt_2j39ngz8ftz40j1mn00000gn/T/ansible_Tj2nE6/ansible_module_nxos_vrf.py", line 467, in <module>
    main()
  File "/var/folders/jv/29pmt_2j39ngz8ftz40j1mn00000gn/T/ansible_Tj2nE6/ansible_module_nxos_vrf.py", line 453, in main
    commands = map_obj_to_commands((want, have), module)
  File "/var/folders/jv/29pmt_2j39ngz8ftz40j1mn00000gn/T/ansible_Tj2nE6/ansible_module_nxos_vrf.py", line 258, in map_obj_to_commands
    if have.get('vni') and have.get('vni') != '':
AttributeError: 'list' object has no attribute 'get'

```

### PR DESCRIPTION
**Pull request by trishnaguha**
Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

##### SUMMARY

Fixes https://github.com/ansible/ansible/issues/36461
Also the module doesn't work when existing definition/property of vrf is attempted to be modified. The PR fixes that.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

modules/network/nxos/nxos_vrf.py
##### ANSIBLE VERSION

```
devel
```

### CODE REVIEW

File: lib/ansible/modules/network/nxos/nxos_vrf.py [hunk 1]
```diff
@@ -242,21 +242,21 @@ def map_obj_to_commands(updates, module):
...
+                    candidate = w.get(item)
+                    if candidate:
+                        cmd = item + ' ' + str(candidate)
+                        commands.append(cmd)
                 if admin_state == 'up':
                     commands.append('no shutdown')
                 elif admin_state == 'down':
                     commands.append('shutdown')
 
                 if commands:
```
Comment (gdpak) [inline lib/ansible/modules/network/nxos/nxos_vrf.py]: Probably not related to this diff but what is the use of this if command when command is already declared so it seems like a  if True type of check.

File: lib/ansible/modules/network/nxos/nxos_vrf.py [hunk 2]
```diff
@@ -224,12 +224,12 @@ def map_obj_to_commands(updates, module):
...
 
+    args = ('rd', 'description', 'vni')
+
     for w in want:
         name = w['name']
-        description = w['description']
-        vni = w['vni']
-        rd = w['rd']
         admin_state = w['admin_state']
+        vni = w['vni']
```
Comment (gdpak) [inline lib/ansible/modules/network/nxos/nxos_vrf.py:232]: This is not required also like description and rd. Anyway you are setting it below from want

File: lib/ansible/modules/network/nxos/nxos_vrf.py [hunk 3]
```diff
@@ -242,21 +242,21 @@ def map_obj_to_commands(updates, module):
...
                 elif admin_state == 'down':
                     commands.append('shutdown')
 
                 if commands:
                     if vni:
-                        if have.get('vni') and have.get('vni') != '':
-                            commands.insert(1, 'no vni {0}'.format(have['vni']))
+                        for h in have:
+                            if h.get('vni'):
+                                commands.insert(1, 'no vni {0}'.format(h['vni']))
```
Comment (gdpak) [inline lib/ansible/modules/network/nxos/nxos_vrf.py]: Can you please add comments why we need to unconfigure vni is some another vrf which is already configured on box ?

File: lib/ansible/modules/network/nxos/nxos_vrf.py (hunk 2 above)
Comment (trishnaguha) [reply lib/ansible/modules/network/nxos/nxos_vrf.py:232]: This is required for L263 `if vni:`

File: lib/ansible/modules/network/nxos/nxos_vrf.py (hunk 3 above)
Comment (trishnaguha) [reply lib/ansible/modules/network/nxos/nxos_vrf.py]: +1

File: lib/ansible/modules/network/nxos/nxos_vrf.py (hunk 1 above)
Comment (trishnaguha) [reply lib/ansible/modules/network/nxos/nxos_vrf.py]: Good point. I will remove this check.

**Review by gdpak — APPROVED** [review]
looks good now

**Pull request merged**